### ISSUE TITLE
xcode test does not upload results to TFS

### ISSUE THREAD

**Issue by kristoferreierson** [Microsoft/vsts-tasks#5972]
## Troubleshooting
Checkout how to troubleshoot failures and collect debug logs: https://docs.microsoft.com/en-us/vsts/build-release/actions/troubleshooting

## Environment
- Server - VSTS or TFS on-premises?

VSTS    

    - If using TFS on-premises, provide the version: 
    
    - If using VSTS, provide the account name, team project name, build definition name/build number: 

msazure.com / Intune / Build 2.0

- Agent - Hosted or Private: 

Hosted
    
    - If using Hosted agent, provide agent queue name:

    - If using private agent, provide the OS of the machine running the agent and the agent version: 

OSX/ 2.124.0

2017-12-04 15:21:35Z INFO AgentProcess] Agent is built for OSX - osx.10.11-x64.
[2017-12-04 15:21:35Z INFO AgentProcess] RuntimeInformation: Darwin 16.7.0 Darwin Kernel Version 16.7.0: Wed Oct  4 00:17:00 PDT 2017; root:xnu-3789.71.6~1/RELEASE_X86_64.
[2017-12-04 15:21:35Z INFO AgentProcess] Version: 2.124.0
[2017-12-04 15:21:35Z INFO AgentProcess] Commit: 4da675de68ad8b21c437d80fd2be9ea762355633
[2017-12-04 15:21:35Z INFO AgentProcess] Culture: en-US
[2017-12-04 15:21:35Z INFO AgentProcess] UI Culture: en-US

## Issue Description

I am running an xcode build job with a test action. I have xcpretty installed, and under Advanced I have 'Use Xcpretty' and 'Publish to VSTS/TFS' enabled. There are two issues:

If the test process fails to complete successfully, you don't publish partial results. Looking at the results here, I'd expect to see some results posted:
2017-12-04T14:22:18.0225300Z Executed 789 tests, with 390 failures (8 unexpected) in 382.600 (384.844) seconds
2017-12-04T14:22:18.0307620Z ##[error]Error: /usr/bin/xcodebuild failed with return code: 65
2017-12-04T14:22:18.0361330Z ##[section]Finishing: Xcode test

Even if the job completes successfully, you don't post the results - I changed to a different test target and I see:
2017-12-04T14:53:46.3429710Z Executed 149 tests, with 0 failures (0 unexpected) in 88.165 (88.710) seconds
2017-12-04T14:53:47.2678620Z ##[error]Unable to process command '##vso[results.publish type=JUnit;publishRunAttachments=true;resultFiles=;]' successfully. Please reference documentaion (http://go.microsoft.com/fwlink/?LinkId=817296)
2017-12-04T14:53:47.2738490Z ##[error]'TestResults' has to be specified.
2017-12-04T14:53:47.4064770Z ##[section]Finishing: Xcode test

3. NIT: in the error string above 'documentation' is missing the t:
successfully. Please reference **documentaion** (http://go.microsoft.com/fwlink/?LinkId=817296)

### Error logs

[Insert error from the logs here]

**Comment by kristoferreierson** [comment]
You are writing the results to $(System.DefaultWorkingDirectory)\build.  My repo has a folder called $(System.DefaultWorkingDirectory)/build.  Can you write/read the results from $(Build.ArtifactStagingDirectory)/TestResults?

**Comment by madhurig** [comment]
@kristoferreierson : What version of TFS are you using? Can you provide the version of the Xcode task?

**Comment by kristoferreierson** [comment]
XCode task: 4.*
Agent: 2.124.0

There is another more serious bug.  If any of the unit tests fail, the task doesn't upload the results to TFS.  Looking at the code, in xcode.ts, my guess is that this:

        //--- Xcode Build ---
        await xcb.exec();

Is throwing an exception because xcodebuild returns a non-zero exit code.  This prevents the block in:
 // Test Publishing 
From running.

If you fix anything, fix this.

**Comment by madhurig** [comment]
@kristoferreierson: I am working on a fix to publish test results when the test run fails. Thanks for reporting this. I am surprised we didn't find it and no one reported it till now. 
 
Regarding the issue of test results not being uploaded on success, we look for test results in the "Working Directory". By default, it is under the sources directory. Can you share the full debug log from your build so I can check why this is not working for you? You can attach here or email it to me -  madhurig at microsoft.com.

Thanks,
Madhuri

**Comment by kristoferreierson** [comment]
Madhuri - I will send them via email.  It's pretty easy to repro - if your repo has a folder off the root called 'Build' (note capital). As a work around, I renamed the folder to BuildTools.

**Comment by madhurig** [comment]
Fix has been merged. It will be deployed to VSTS in January 2018 after the holidays.

**Comment by madhurig** [comment]
Fix was deployed to all VSTS accounts in January. Let us now if you still see issues.

**Comment by sakthia01** [comment]
Hi Team,
In VSTS, I am using Xcode test in the build definition.

All my functional test case and UI Test case are executed successfully. Now where I can get the results (Code coverage).

Please find my build definition, If I selected both the checkbox "Use XCPretty and Publish test result to VSTS/TFS" I am getting a warning your test results are not formatted to JUnit.

If it executing the test results where I can find the results.
![screen shot 2018-07-12 at 11 40 20 am](https://user-images.githubusercontent.com/40980145/42615748-2f29a348-85c9-11e8-8243-931ed43a09bf.png)

**Comment by sakthia01** [comment]
Hi Madhuri,

I am running my Xcode project in Mac OS X build agent.
My code coverage details are under,

/Library/Developer/Xcode/DerivedData/AppName-athwtzejaaupathktnqcilwumvln/Logs/Test/xxxxxx-xxxx-xxxx-xxxx-xxxxxxxx.xccovreport

In build agent machine.

Now, how do I display this code coverage report in VSTS?
Which tool will help me to display this coverage value?

**Comment by madhurig** [comment]
@sakthia01: We currently don't support uploading code coverage results from the Xcode task. The Xcode CC results will need to be converted to another format and then uploaded. You can look into other tools like [slather](https://github.com/SlatherOrg/slather) and script that in your VSTS build process. 

I have added it to our backlog to provide more in the box support for code coverage with the VSTS Xcode task.

Thanks,
Madhuri Gummalla
Engineering Manager
VSTS/TFS

**Comment by christoff-1992** [comment]
@madhurig - It would appear that this is still an issue, if the tests fail then the test's tab in Azure DevOps displays. `'No test runs are available for this build. Enable automated tests in your build pipeline by running your test framework of choice, such as JUnit, Visual Studio Test, or xUnit. If you choose to run tests using a custom task or runner, you can publish results using the Publish Test Results task.'`

If I can assist in providing any logs please let me know.

**Comment by madhurig** [comment]
@christoff-1992 - are you using xcpretty? Test results are only published if xcpretty is enabled.

**Comment by christoff-1992** [comment]
@madhurig - Yes I am using xcpretty build the results are not been published.

This is the configuration that I am using for the task.

```yml
- task: Xcode@5
  displayName: 'Test the XXXXXXXXXX project'
  inputs:
    actions: 'test'
    scheme: '$(scheme)'
    sdk: '$(sdk)'
    configuration: '$(configuration)'
    xcWorkspacePath: '$(workspace)'
    xcodeVersion: '$(xcode)'
    destinationSimulators: 'iPhone X'
    destinationTypeOption: 'simulators'
    destinationPlatformOption: 'iOS'
    signingOption: "auto"
    args: '-UseModernBuildSystem=YES'
    useXcpretty: true
    publishJUnitResults: true
    packageApp: false
```

The only thing I can think of is that the tests are failing with an exit code of `65`, and this may be due to running the tests in parallel. It looks like the simulators may be shutting down early causing the tests to stop executing and return the `exit code 65`. The particular tests that are showing as failed in the build logs do not have any reason and hence why I wanted to check the results.

**Comment by madhurig** [comment]
@christoff-1992 - Tests running twice might be an xcodebuild setting. When run in parallel, my hunch is that the junit format results file is not getting written by xcpretty, so nothing is getting published to the server. Please open a new issue.

**Comment by tumatauenga** [comment]
> @sakthia01: We currently don't support uploading code coverage results from the Xcode task. The Xcode CC results will need to be converted to another format and then uploaded. You can look into other tools like [slather](https://github.com/SlatherOrg/slather) and script that in your VSTS build process.
> 
> I have added it to our backlog to provide more in the box support for code coverage with the VSTS Xcode task.
> 
> Thanks,
> Madhuri Gummalla
> Engineering Manager
> VSTS/TFS

Any update on supporting code coverage results from the Xcode task? Are you able to provide any help with regards to getting Slather working to do it instead?
